phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java

```diff
@@ -357,20 +357,33 @@ RegionScanner getWrappedScanner(final ObserverContext<RegionCoprocessorEnvironme
                 dataRegion, indexMaintainer, null, viewConstants, null, null, projector, ptr, useQualiferAsListIndex);
     }
 
-    public void setScanOptionsForFlushesAndCompactions(ScanOptions options) {
+    public void setScanOptionsForFlushesAndCompactions(Store store, ScanOptions options,
+            boolean retainAllVersions) {
         // We want the store to give us all the deleted cells to StoreCompactionScanner
         options.setKeepDeletedCells(KeepDeletedCells.TTL);
         options.setTTL(HConstants.FOREVER);
-        options.setMaxVersions(Integer.MAX_VALUE);
-        options.setMinVersions(Integer.MAX_VALUE);
+        if (retainAllVersions) {
+            options.setMaxVersions(Integer.MAX_VALUE);
+            options.setMinVersions(Integer.MAX_VALUE);
+        } else {
+            options.setMinVersions(Math.max(Math.max(options.getMaxVersions(),
+                    store.getColumnFamilyDescriptor().getMaxVersions()), 1));
+            options.setMinVersions(Math.max(Math.max(options.getMinVersions(),
+                    store.getColumnFamilyDescriptor().getMaxVersions()), 1));
```

Comment:
is min versions supposed to be columnFamilyDescription max versions if it exists, or should this also be min? `store.getColumnFamilyDescriptor().getMinVersions()`

Reply:
This code is borrowed from the previous implementation (before CompactionScanner).  It is interesting that some tests for old indexes or SYCAT fails if getMaxVersions is not used above. I do not know the exact reason for it. This will be changed very likely again when we enable CompactionScanner for minor compaction. I am planning to get that change (for https://issues.apache.org/jira/browse/PHOENIX-7314) soon.

Reply:
Yeah I could see how potentially we still want to retain all versions here for whatever reason... I did not try to understand it. But on line 369 above we are also calling `options.setMinVersions` so we call `options.setMinVersions` twice, but never call `options.setMaxVersions`. If the CFDescriptor can have a higher number for maxVersions than what `options` contains, then min will be greater than max.

Reply:
I just noticed that there is a typo here. The old code just had `options.setMinVersions(Math.max(Math.max(options.getMinVersions(),
                    store.getColumnFamilyDescriptor().getMaxVersions()), 1));`.  I wanted to have the same logic for max version and even though it was not necessary for existing tests. So my intention was to add `options.setMaxVersions(Math.max(Math.max(options.getMaxVersions(),
                    store.getColumnFamilyDescriptor().getMaxVersions()), 1));` However, because of the typo, I ended up not adding this logic for max version. The code is harmless but confusing. I am going to change this in the jira. Thank you for catching this.